Frameworks/UIKit/NSLayoutManager.mm

```diff
@@ -22,213 +22,163 @@
-            }
+    for (id lineRef in static_cast<NSArray*>(CTFrameGetLines(_frame))) {
+        CTLineRef line = static_cast<CTLineRef>(lineRef);
 
-            //  Did we determine that we can fit at least one character?
-            if (pos != curIdx) {
-                //  Yes, we will record what the (x,y) origin is and create a new CTLine from it
-                CFRange lineRange;
-                lineRange.location = curIdx;
-                lineRange.length = pos - curIdx;
+        // Width of line already drawn, saves us from having to redraw line twice
+        double drawnWidth = 0.0;
 
-                CTLineRef line = CTTypesetterCreateLine(ts, lineRange);
+        CGFloat ascent, descent, leading;
 
-                float ascent = 0.0f, descent = 0.0f, leading = 0.0f;
-                float width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
-                float lineHeight = ascent - descent + leading;
+        // Width of entire line
+        double width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
 
-                CGPoint lineOrigin;
-                lineOrigin = startPoint;
-                [_ctLines addObject:(id)line];
-                _lineOrigins.push_back(lineOrigin);
-                CFRelease(line);
+        // Maximum height of lines on current horizontal, needed to get next yPos
+        CGFloat lineHeight = leading + ascent - descent;
 
-                //  Record what the height of the line was; we'll need to adjust all lines that fit within
-                curIdx = pos;
-                if (lineOrigin.x + width > _totalSize.width) {
-                    _totalSize.width = lineOrigin.x + width;
-                }
-                if (lineHeight > maxLineHeight) {
-                    maxLineHeight = lineHeight;
-                }
-                if (ascent > maxAscent) {
-                    maxAscent = ascent;
-                }
+        CFRange lineRange = CTLineGetStringRange(line);
+
+        // Index of first glyph in line to be drawn
+        CFIndex stringIndex = lineRange.location;
+        CFIndex lineEnd = stringIndex + lineRange.length;
+
+        while (stringIndex < lineEnd) {
+            if (origin.y > containerSize.height) {
+                // Added as much text as can fit in the frame, can end here
+                return;
             }
 
-            //  If the last character was a line break, exit this line early
-            if (curIdx > 0) {
-                unichar c = [string characterAtIndex:curIdx - 1];
-                if (c == 10 || c == 13) {
-                    lastRunWasLineBreak = true;
-                    break;
-                }
+            if (!__lineHasGlyphsAfterIndex(line, stringIndex)) {
+                // Ended up with white space at end of line, break to next line
+                origin = { 0.0f, origin.y + lineHeight };
+                _totalSize.height += lineHeight;
+                break;
             }
```

Comment:
Noooooooooo.... we are going back to interpreting and laying out text implementation ourself sort of design again here. We should not be doing this. Though it might all seem to work when you bring in non en-us languages some of these assumptions is going to break.